File: frontend/src/app/core/services/asistencia.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

// ── Modelos / Interfaces ───────────────────────────────────────────────────

export type EstadoAsistencia =
  | 'ASISTIO'
  | 'FALTA'
  | 'TARDANZA'
  | 'LICENCIA'
  | 'PERMISO_ACADEMIA'
  | 'JUSTIFICADO';

export interface RegistroAsistenciaDto {
  alumnoId: number;
  estado: EstadoAsistencia;
  justificacion?: string;
}

export interface ResumenAsistencia {
  totalDias: number;
  asistencias: number;
  faltas: number;
  tardanzas: number;
  justificados: number;
  permisosAcademia: number;
  porcentajeAsistencia: number;
}

export interface Asistencia {
  id: number;
  alumno: { id: number; nombreCompleto: string };
  aula: { id: number; descripcion: string };
  fecha: string;
  estado: EstadoAsistencia;
  horaLlegada?: string;
  minutosTardanza?: number;
  aplicadoPermisoAcademia: boolean;
  horaPermisoAcademia?: string;
  justificacion?: string;
  tieneJustificacion: boolean;
  sincronizadoOffline: boolean;
}

export interface AlertaFalta {
  alumnoId: number;
  nombreAlumno: string;
  totalFaltas: number;
}

export interface PermisoAcademia {
  tienePermiso: boolean;
  horaEntrada: string;
}

// ── Servicio ───────────────────────────────────────────────────────────────

/**
 * AsistenciaService – Servicio Angular para conectar con el backend Spring Boot.
 *
 * Implementa:
 *  - Registro individual y por lote (app móvil docente)
 *  - Sincronización de registros offline
 *  - Resumen e historial del alumno (Portal del Alumno – Pestaña 2)
 *  - Alertas de faltas excesivas (Panel Director)
 *  - Verificación de permiso de academia
 */
@Injectable({
  providedIn: 'root'
})
export class AsistenciaService {

  private readonly BASE_URL = `${environment.apiUrl}/asistencias`;

  constructor(private http: HttpClient) {}

  // ── Registro ──────────────────────────────────────────────────────────

  /**
   * Registra la asistencia de un alumno individual.
   */
  registrarAsistencia(
    alumnoId: number,
    fecha: string,          // formato: YYYY-MM-DD
    estado: EstadoAsistencia,
    docenteId: number
  ): Observable<Asistencia> {
    const params = new HttpParams()
      .set('alumnoId', alumnoId)
      .set('fecha', fecha)
      .set('estado', estado)
      .set('docenteId', docenteId);

    return this.http.post<Asistencia>(`${this.BASE_URL}/registro`, null, { params });
  }

  /**
   * Registro masivo de asistencia para un aula completa.
   * Optimizado para la vista móvil de marcado rápido del docente.
   */
  registrarAsistenciaLote(
    aulaId: number,
    fecha: string,
    docenteId: number,
    registros: RegistroAsistenciaDto[]
  ): Observable<Asistencia[]> {
    const params = new HttpParams()
      .set('aulaId', aulaId)
      .set('fecha', fecha)
      .set('docenteId', docenteId);

    return this.http.post<Asistencia[]>(`${this.BASE_URL}/lote`, registros, { params });
  }

  // ── Consultas ─────────────────────────────────────────────────────────

  /**
   * Obtiene las asistencias de un aula para la fecha dada (tiempo real).
   */
  obtenerAsistenciaAula(aulaId: number, fecha: string): Observable<Asistencia[]> {
    return this.http.get<Asistencia[]>(`${this.BASE_URL}/aula/${aulaId}`, {
      params: new HttpParams().set('fecha', fecha)
    });
  }

  /**
   * Resumen de asistencia del alumno para el Portal del Alumno (Pestaña 2 – Calendario).
   */
  obtenerResumenAlumno(
    alumnoId: number,
    inicio: string,
    fin: string
  ): Observable<ResumenAsistencia> {
    const params = new HttpParams().set('inicio', inicio).set('fin', fin);
    return this.http.get<ResumenAsistencia>(
      `${this.BASE_URL}/alumno/${alumnoId}/resumen`, { params }
    );
  }

  /**
   * Historial detallado de asistencias del alumno (para el calendario).
   */
  obtenerHistorialAlumno(
    alumnoId: number,
    inicio: string,
    fin: string
  ): Observable<Asistencia[]> {
    const params = new HttpParams().set('inicio', inicio).set('fin', fin);
    return this.http.get<Asistencia[]>(
      `${this.BASE_URL}/alumno/${alumnoId}/historial`, { params }
    );
  }

  // ── Offline Sync ──────────────────────────────────────────────────────

  /**
   * Sincroniza registros de asistencia capturados offline en el dispositivo del docente.
   * Los registros se guardan en localStorage y se sincronizan cuando hay conexión.
   */
  sincronizarRegistrosOffline(registros: Asistencia[]): Observable<{ sincronizados: number }> {
    return this.http.post<{ sincronizados: number }>(
      `${this.BASE_URL}/offline/sync`, registros
    );
  }

  // ── Permiso de Academia ───────────────────────────────────────────────

  /**
   * Verifica si el alumno tiene permiso de academia (entrada 13:30 / 14:30).
   */
  verificarPermisoAcademia(alumnoId: number): Observable<PermisoAcademia> {
    return this.http.get<PermisoAcademia>(
      `${this.BASE_URL}/alumno/${alumnoId}/permiso-academia`
    );
  }

  // ── Alertas (Panel Director) ──────────────────────────────────────────

  /**
   * Alumnos con faltas excesivas en el período (Monitor del Director).
   */
  obtenerAlertasFaltasExcesivas(
    aulaId: number,
    inicio: string,
    fin: string,
    minFaltas = 3
  ): Observable<AlertaFalta[]> {
    const params = new HttpParams()
      .set('aulaId', aulaId)
      .set('inicio', inicio)
      .set('fin', fin)
      .set('minFaltas', minFaltas);

    return this.http.get<AlertaFalta[]>(`${this.BASE_URL}/alertas/faltas-excesivas`, { params });
  }

  // ── Persistencia Offline Local ────────────────────────────────────────

  /**
   * Guarda registros offline en localStorage para sincronizar luego.
   */
  guardarOfflineLocal(asistencias: RegistroAsistenciaDto[]): void {
    const key = 'bastidas_asistencias_offline';
    const existentes = this.cargarOfflineLocal();
    const combinados = [...existentes, ...asistencias];
    localStorage.setItem(key, JSON.stringify(combinados));
  }

  cargarOfflineLocal(): RegistroAsistenciaDto[] {
    const raw = localStorage.getItem('bastidas_asistencias_offline');
    return raw ? JSON.parse(raw) : [];
  }

  limpiarOfflineLocal(): void {
    localStorage.removeItem('bastidas_asistencias_offline');
  }

  hayRegistrosPendientesOffline(): boolean {
    return this.cargarOfflineLocal().length > 0;
  }
}

File: backend/src/main/java/com/colegio/bastidas/service/AsistenciaService.java
package com.colegio.bastidas.service;

import com.colegio.bastidas.model.Asistencia;
import com.colegio.bastidas.model.Asistencia.EstadoAsistencia;

import java.time.LocalDate;
import java.util.List;

/**
 * Contrato de negocio para el Control de Asistencia.
 * Incluye validaciones para el Permiso de Academia y alertas de inasistencia.
 */
public interface AsistenciaService {

    /**
     * Registra la asistencia de un alumno para la fecha indicada.
     * Aplica automáticamente la lógica de permiso de academia si corresponde.
     *
     * @param alumnoId ID del alumno
     * @param fecha    Fecha del registro
     * @param estado   Estado de asistencia observado
     * @param docenteId ID del docente que registra
     * @return Asistencia persistida
     */
    Asistencia registrarAsistencia(Long alumnoId, LocalDate fecha,
                                   EstadoAsistencia estado, Long docenteId);

    /**
     * Registra en lote la asistencia de toda un aula para una fecha.
     * Optimizado para la vista móvil de marcado rápido.
     *
     * @param aulaId    ID del aula
     * @param fecha     Fecha del registro
     * @param registros Lista de pares (alumnoId → estado)
     * @param docenteId Docente que registra
     * @return Lista de asistencias persistidas
     */
    List<Asistencia> registrarAsistenciaLote(Long aulaId, LocalDate fecha,
                                              List<RegistroAsistenciaDto> registros,
                                              Long docenteId);

    /**
     * Verifica si un alumno tiene permiso de academia activo y cuál es
     * su horario de entrada tolerado (13:30 o 14:30).
     *
     * @param alumnoId ID del alumno
     * @return true si tiene permiso de academia activo
     */
    boolean verificarHorarioAcademia(Long alumnoId);

    /**
     * Obtiene la hora de entrada permitida para un alumno con permiso de academia.
     *
     * @param alumnoId ID del alumno
     * @return String con la hora "13:30" o "14:30", o null si no tiene permiso
     */
    String obtenerHoraPermisoAcademia(Long alumnoId);

    /**
     * Procesa y envía alertas de inasistencia al apoderado vía notificación.
     * Se ejecuta automáticamente cuando un alumno acumula N faltas en el período.
     *
     * @param alumnoId   ID del alumno
     * @param fechaInicio Inicio del período de análisis
     * @param fechaFin    Fin del período de análisis
     */
    void procesarAlertaInasistencia(Long alumnoId, LocalDate fechaInicio, LocalDate fechaFin);

    /**
     * Sincroniza registros de asistencia capturados offline en el dispositivo móvil.
     * Aplica deduplicación para evitar duplicados.
     *
     * @param registros Lista de asistencias capturadas offline
     * @return Número de registros sincronizados exitosamente
     */
    int sincronizarRegistrosOffline(List<Asistencia> registros);

    /**
     * Obtiene el resumen de asistencia de un alumno en un rango de fechas.
     *
     * @param alumnoId ID del alumno
     * @param inicio   Fecha de inicio
     * @param fin      Fecha de fin
     * @return Resumen con totales por estado
     */
    ResumenAsistenciaDto obtenerResumenAlumno(Long alumnoId, LocalDate inicio, LocalDate fin);

    /**
     * Obtiene la lista de asistencias del aula para una fecha específica.
     *
     * @param aulaId ID del aula
     * @param fecha  Fecha a consultar
     * @return Lista de asistencias del aula
     */
    List<Asistencia> obtenerAsistenciasAulaPorFecha(Long aulaId, LocalDate fecha);

    /**
     * Detecta alumnos con faltas excesivas en el período para el panel del director.
     *
     * @param aulaId   ID del aula
     * @param inicio   Inicio del período
     * @param fin      Fin del período
     * @param minFaltas Umbral mínimo de faltas para la alerta
     * @return Lista de alumnos con faltas excesivas
     */
    List<AlertaFaltaDto> detectarAlumnosConFaltasExcesivas(Long aulaId,
                                                            LocalDate inicio,
                                                            LocalDate fin,
                                                            int minFaltas);

    // ── DTOs internos del servicio ─────────────────────────────────────────
    record RegistroAsistenciaDto(Long alumnoId, EstadoAsistencia estado, String justificacion) {}

    record ResumenAsistenciaDto(
        int totalDias,
        int asistencias,
        int faltas,
        int tardanzas,
        int justificados,
        int permisosAcademia,
        double porcentajeAsistencia
    ) {}

    record AlertaFaltaDto(Long alumnoId, String nombreAlumno, long totalFaltas) {}
}
